File: mini-cursor/minicursor/checkpoints.py
"""File-level checkpoints for /undo.

Before `write_file`/`edit_file` mutate a path, the agent records that path's
pre-turn content here. When the turn ends, everything recorded during it is
persisted as one checkpoint (a JSON file under `~/.minicursor/checkpoints/`,
keyed by a hash of the workspace path so unrelated projects don't collide).
`/undo` pops the most recent checkpoint and restores every file it touched to
its pre-turn state (or deletes it, if the file didn't exist before the turn).

Scope, on purpose: this only covers `write_file`/`edit_file`, the two tools
mini-cursor fully controls and already diffs before applying. `run_command`
can do anything (install packages, mutate unrelated state, talk to a
network) — there is no honest way to snapshot and revert that generically, so
it isn't covered. Checkpoints are a safety net for agent-authored file edits,
not a full VM undo.
"""

from __future__ import annotations

import hashlib
import json
import time
import uuid
from dataclasses import dataclass, field
from pathlib import Path

CHECKPOINTS_ROOT = Path.home() / ".minicursor" / "checkpoints"

# Skip checkpointing files above this size (still allowed to be written, just
# not covered by /undo) so a single huge file can't blow up disk usage.
MAX_CHECKPOINT_FILE_BYTES = 5_000_000

# Retention cap per workspace so checkpoints don't accumulate forever.
MAX_CHECKPOINTS_PER_WORKSPACE = 50


@dataclass
class FileChange:
    path: str  # absolute path, resolved inside the workspace it was recorded for
    existed_before: bool
    before_content: str | None  # None means the file did not exist before this turn

    def to_json(self) -> dict:
        return {
            "path": self.path,
            "existed_before": self.existed_before,
            "before_content": self.before_content,
        }

    @staticmethod
    def from_json(d: dict) -> "FileChange":
        return FileChange(
            path=d["path"],
            existed_before=bool(d["existed_before"]),
            before_content=d.get("before_content"),
        )


@dataclass
class Checkpoint:
    id: str
    created_at: float
    label: str
    changes: list[FileChange] = field(default_factory=list)
    skipped_large: list[str] = field(default_factory=list)
    skipped_binary: list[str] = field(default_factory=list)

    def to_json(self) -> dict:
        return {
            "id": self.id,
            "created_at": self.created_at,
            "label": self.label,
            "changes": [c.to_json() for c in self.changes],
            "skipped_large": self.skipped_large,
            "skipped_binary": self.skipped_binary,
        }

    @staticmethod
    def from_json(d: dict) -> "Checkpoint":
        return Checkpoint(
            id=str(d["id"]),
            created_at=float(d["created_at"]),
            label=str(d.get("label", "")),
            changes=[FileChange.from_json(c) for c in d.get("changes", [])],
            skipped_large=list(d.get("skipped_large", [])),
            skipped_binary=list(d.get("skipped_binary", [])),
        )


def _workspace_dir(root: Path) -> Path:
    key = hashlib.sha256(str(root.resolve()).encode("utf-8")).hexdigest()[:16]
    return CHECKPOINTS_ROOT / key


class CheckpointStore:
    """Per-workspace stack of checkpoints, persisted so /undo survives restarts.

    Nothing is written to disk (not even an empty directory) until a turn
    actually mutates a file — merely opening a workspace shouldn't leave
    traces in ~/.minicursor.
    """

    def __init__(self, workspace_root: Path, storage_dir: Path | None = None):
        self.root = workspace_root.resolve()
        self.dir = storage_dir if storage_dir is not None else _workspace_dir(self.root)
        self._current: Checkpoint | None = None

    # -- recording (during a turn) ---------------------------------------

    def begin_turn(self, label: str) -> None:
        """Start collecting changes for a new turn. Call once per user turn."""
        self._current = Checkpoint(
            id=uuid.uuid4().hex[:12],
            created_at=time.time(),
            label=" ".join(label.split())[:120],
        )

    def record_before(self, target: Path) -> None:
        """Capture `target`'s current (pre-mutation) content, once per turn per path.

        Only ever called for paths write_file/edit_file are *about* to touch —
        the actual write/edit may still go on to fail its own validation (e.g.
        `target` turns out to be a directory, or an edit's old_string doesn't
        match). This deliberately records nothing in that case rather than a
        checkpoint entry for a change that never happened.
        """
        if self._current is None:
            return
        key = str(target)
        if any(c.path == key for c in self._current.changes):
            return  # keep the *first* pre-turn state if the same file is touched twice
        if target.exists() and not target.is_file():
            return  # not a regular file (e.g. a directory) — write_file/edit_file will
            # reject this on their own; nothing meaningful to checkpoint or undo.
        existed = target.is_file()
        before: str | None
        if existed:
            try:
                if target.stat().st_size > MAX_CHECKPOINT_FILE_BYTES:
                    if key not in self._current.skipped_large:
                        self._current.skipped_large.append(key)
                    return
                raw = target.read_bytes()
            except OSError:
                return
            try:
                before = raw.decode("utf-8")
            except UnicodeDecodeError:
                # Not valid UTF-8: read_text(errors="replace") would silently
                # corrupt the byte content before it's even stored, making
                # /undo "restore" a lossy, already-mangled version. Refuse to
                # checkpoint it instead — same honest-skip treatment as an
                # oversized file.
                if key not in self._current.skipped_binary:
                    self._current.skipped_binary.append(key)
                return
        else:
            before = None
        self._current.changes.append(FileChange(path=key, existed_before=existed, before_content=before))

    def commit_turn(self) -> Checkpoint | None:
        """Persist the turn's recorded changes as one checkpoint. Returns it, or
        None if the turn recorded nothing at all (no file was touched).

        A turn that touched only skipped (oversized/binary) files still comes
        back non-None, so the caller can tell the user those files aren't
        covered by /undo — but nothing is written to disk for it, since there
        is no actual undo-able state to persist or list in /checkpoints.
        """
        cp, self._current = self._current, None
        if cp is None or (not cp.changes and not cp.skipped_large and not cp.skipped_binary):
            return None
        if not cp.changes:
            return cp  # nothing to persist — just skipped-file bookkeeping to report
        self.dir.mkdir(parents=True, exist_ok=True)
        # A monotonic sequence number (not just the wall-clock timestamp) so
        # ordering stays correct even when two turns commit within the same
        # clock tick — filename sort must match commit order for /undo to
        # ever pop the right one.
        seq = self._next_sequence()
        path = self.dir / f"{seq:010d}_{cp.id}.json"
        path.write_text(json.dumps(cp.to_json(), ensure_ascii=False, indent=2), encoding="utf-8")
        self._prune()
        return cp

    def _next_sequence(self) -> int:
        max_seq = -1
        for f in self._files():
            try:
                max_seq = max(max_seq, int(f.name.split("_", 1)[0]))
            except ValueError:
                continue
        return max_seq + 1

    # -- listing / undo ---------------------------------------------------

    def _files(self) -> list[Path]:
        if not self.dir.is_dir():
            return []
        return sorted(self.dir.glob("*.json"))

    def _prune(self) -> None:
        files = self._files()
        excess = len(files) - MAX_CHECKPOINTS_PER_WORKSPACE
        for f in files[: max(excess, 0)]:
            f.unlink(missing_ok=True)

    def count(self) -> int:
        """Total number of checkpoints retained on disk (may exceed what list() returns)."""
        return len(self._files())

    def list(self, limit: int = 10) -> list[Checkpoint]:
        """Most recent first."""
        if limit <= 0:
            return []
        out = []
        for f in reversed(self._files()[-limit:]):
            try:
                out.append(Checkpoint.from_json(json.loads(f.read_text(encoding="utf-8"))))
            except (OSError, json.JSONDecodeError, KeyError, TypeError, ValueError):
                continue
        return out

    def undo_last(self) -> Checkpoint | None:
        """Pop and apply the most recent checkpoint. Returns it, or None if empty.

        Best-effort per file: a problem restoring one change (permission
        error, a path that no longer makes sense, a symlink now pointing
        somewhere unexpected) doesn't abort the rest of the checkpoint or
        crash the caller — every apply below and the checkpoint file itself
        are always cleaned up.
        """
        files = self._files()
        if not files:
            return None
        last = files[-1]
        try:
            cp = Checkpoint.from_json(json.loads(last.read_text(encoding="utf-8")))
        except (OSError, json.JSONDecodeError, KeyError, TypeError, ValueError):
            last.unlink(missing_ok=True)
            return None
        for change in cp.changes:
            try:
                # Re-resolve at undo time (not just the raw stored path) so a
                # symlink introduced after the checkpoint was recorded can't
                # smuggle the write outside the workspace.
                target = Path(change.path).resolve()
                if target != self.root and self.root not in target.parents:
                    continue  # refuse to touch anything outside this checkpoint's workspace
                if change.existed_before:
                    target.parent.mkdir(parents=True, exist_ok=True)
                    target.write_text(change.before_content or "", encoding="utf-8")
                elif target.is_file():
                    target.unlink()
            except OSError:
                continue  # best-effort: skip this file, still process the rest
        last.unlink(missing_ok=True)
        return cp

File: tythan-code/src/core/checkpoints.ts
/**
 * File-level checkpoints for undo.
 *
 * Before write_file/edit_file mutate a path, the agent records that path's
 * pre-turn content here. When the turn ends, everything recorded during it
 * is persisted as one checkpoint (a JSON file under a caller-supplied
 * storage directory). Undo pops the most recent checkpoint and restores
 * every file it touched to its pre-turn state (or deletes it, if the file
 * didn't exist before the turn).
 *
 * Scope, on purpose: this only covers write_file/edit_file, the two tools
 * Tythan Code fully controls and already diffs before applying. run_command
 * can do anything — there is no honest way to snapshot and revert that
 * generically, so it isn't covered. This is a safety net for agent-authored
 * file edits, not a full VM undo.
 *
 * This module intentionally carries forward every fix found reviewing the
 * original Python port: never checkpoint a non-regular-file target (avoids
 * an EISDIR crash on undo), skip (and honestly report) files that aren't
 * valid UTF-8 instead of silently corrupting them, use a monotonic sequence
 * number rather than a wall-clock timestamp for ordering, and re-resolve
 * paths against the live filesystem at undo time so a symlink introduced
 * after the checkpoint can't smuggle a write outside the workspace.
 */

import * as crypto from "node:crypto";
import * as fs from "node:fs";
import * as path from "node:path";

import { realpathAsFarAsPossible } from "./tools";

// Skip checkpointing files above this size (still allowed to be written,
// just not covered by undo) so a single huge file can't blow up disk usage.
export const MAX_CHECKPOINT_FILE_BYTES = 5_000_000;

// Retention cap per workspace so checkpoints don't accumulate forever.
export const MAX_CHECKPOINTS_PER_WORKSPACE = 50;

export interface FileChange {
  /** absolute path, resolved inside the workspace it was recorded for */
  path: string;
  existedBefore: boolean;
  /** undefined means the file did not exist before this turn */
  beforeContent: string | undefined;
}

export interface Checkpoint {
  id: string;
  createdAt: number;
  label: string;
  changes: FileChange[];
  skippedLarge: string[];
  skippedBinary: string[];
}

function isPlainObject(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null;
}

function fileChangeFromJson(raw: unknown): FileChange | null {
  if (!isPlainObject(raw) || typeof raw.path !== "string") {
    return null;
  }
  return {
    path: raw.path,
    existedBefore: Boolean(raw.existedBefore),
    beforeContent: typeof raw.beforeContent === "string" ? raw.beforeContent : undefined,
  };
}

function checkpointFromJson(raw: unknown): Checkpoint | null {
  if (!isPlainObject(raw)) {
    return null;
  }
  const r = raw;
  if (typeof r.id !== "string" && typeof r.id !== "number") {
    return null;
  }
  const createdAt = typeof r.createdAt === "number" ? r.createdAt : Number(r.createdAt);
  if (!Number.isFinite(createdAt)) {
    return null;
  }
  const changes = Array.isArray(r.changes)
    ? (r.changes.map(fileChangeFromJson).filter((c): c is FileChange => c !== null))
    : [];
  return {
    id: String(r.id),
    createdAt,
    label: typeof r.label === "string" ? r.label : "",
    changes,
    skippedLarge: Array.isArray(r.skippedLarge) ? r.skippedLarge.filter((x): x is string => typeof x === "string") : [],
    skippedBinary: Array.isArray(r.skippedBinary)
      ? r.skippedBinary.filter((x): x is string => typeof x === "string")
      : [],
  };
}

function toJson(cp: Checkpoint): object {
  return {
    id: cp.id,
    createdAt: cp.createdAt,
    label: cp.label,
    changes: cp.changes.map((c) => ({
      path: c.path,
      existedBefore: c.existedBefore,
      beforeContent: c.beforeContent ?? null,
    })),
    skippedLarge: cp.skippedLarge,
    skippedBinary: cp.skippedBinary,
  };
}

/** Stable per-workspace storage key, so unrelated projects don't collide
 * when the caller derives a directory from `CHECKPOINTS_ROOT/<key>`. */
export function workspaceStorageKey(workspaceRoot: string): string {
  return crypto.createHash("sha256").update(path.resolve(workspaceRoot)).digest("hex").slice(0, 16);
}

function isValidUtf8(buf: Buffer): string | null {
  try {
    // { fatal: true } makes TextDecoder throw on invalid byte sequences,
    // unlike Buffer#toString("utf-8") which silently replaces them — exactly
    // the strict check needed to detect a non-UTF-8 file before ever storing
    // a lossy copy of it.
    return new TextDecoder("utf-8", { fatal: true }).decode(buf);
  } catch {
    return null;
  }
}

export class CheckpointStore {
  readonly root: string;
  readonly dir: string;
  private current: Checkpoint | null = null;

  constructor(workspaceRoot: string, storageDir: string) {
    this.root = realpathAsFarAsPossible(path.resolve(workspaceRoot));
    this.dir = storageDir;
  }

  /** Start collecting changes for a new turn. Call once per user turn. */
  beginTurn(label: string): void {
    this.current = {
      id: crypto.randomUUID().replace(/-/g, "").slice(0, 12),
      createdAt: Date.now() / 1000,
      label: label.split(/\s+/).filter(Boolean).join(" ").slice(0, 120),
      changes: [],
      skippedLarge: [],
      skippedBinary: [],
    };
  }

  /** Capture `target`'s current (pre-mutation) content, once per turn per path.
   *
   * Only ever called for paths write_file/edit_file are *about* to touch —
   * the actual write/edit may still go on to fail its own validation (e.g.
   * `target` turns out to be a directory, or an edit's oldString doesn't
   * match). This deliberately records nothing in that case rather than a
   * checkpoint entry for a change that never happened.
   */
  recordBefore(target: string): void {
    if (this.current === null) {
      return;
    }
    const key = target;
    if (this.current.changes.some((c) => c.path === key)) {
      return; // keep the *first* pre-turn state if the same file is touched twice
    }
    let stat: fs.Stats | undefined;
    try {
      stat = fs.statSync(target);
    } catch {
      stat = undefined;
    }
    if (stat && !stat.isFile()) {
      return; // not a regular file (e.g. a directory) — nothing meaningful to checkpoint
    }
    let beforeContent: string | undefined;
    if (stat) {
      if (stat.size > MAX_CHECKPOINT_FILE_BYTES) {
        if (!this.current.skippedLarge.includes(key)) {
          this.current.skippedLarge.push(key);
        }
        return;
      }
      let raw: Buffer;
      try {
        raw = fs.readFileSync(target);
      } catch {
        return;
      }
      const decoded = isValidUtf8(raw);
      if (decoded === null) {
        // Not valid UTF-8: a lossy decode would silently corrupt the byte
        // content before it's even stored, making undo "restore" an
        // already-mangled version. Refuse to checkpoint it instead — same
        // honest-skip treatment as an oversized file.
        if (!this.current.skippedBinary.includes(key)) {
          this.current.skippedBinary.push(key);
        }
        return;
      }
      beforeContent = decoded;
    } else {
      beforeContent = undefined;
    }
    this.current.changes.push({ path: key, existedBefore: stat !== undefined, beforeContent });
  }

  /** Persist the turn's recorded changes as one checkpoint. Returns it, or
   * undefined if the turn recorded nothing at all (no file was touched).
   *
   * A turn that touched only skipped (oversized/binary) files still comes
   * back defined, so the caller can tell the user those files aren't
   * covered by undo — but nothing is written to disk for it, since there is
   * no actual undo-able state to persist or list.
   */
  commitTurn(): Checkpoint | undefined {
    const cp = this.current;
    this.current = null;
    if (cp === null) {
      return undefined;
    }
    if (cp.changes.length === 0 && cp.skippedLarge.length === 0 && cp.skippedBinary.length === 0) {
      return undefined;
    }
    if (cp.changes.length === 0) {
      return cp; // nothing to persist — just skipped-file bookkeeping to report
    }
    fs.mkdirSync(this.dir, { recursive: true });
    // A monotonic sequence number (not just the wall-clock timestamp) so
    // ordering stays correct even when two turns commit within the same
    // clock tick — filename sort must match commit order for undo to ever
    // pop the right one.
    const seq = this.nextSequence();
    const filePath = path.join(this.dir, `${String(seq).padStart(10, "0")}_${cp.id}.json`);
    fs.writeFileSync(filePath, JSON.stringify(toJson(cp), null, 2), "utf-8");
    this.prune();
    return cp;
  }

  private nextSequence(): number {
    let maxSeq = -1;
    for (const f of this.files()) {
      const parsed = parseInt(path.basename(f).split("_", 1)[0] ?? "", 10);
      if (Number.isFinite(parsed)) {
        maxSeq = Math.max(maxSeq, parsed);
      }
    }
    return maxSeq + 1;
  }

  private files(): string[] {
    if (!fs.existsSync(this.dir) || !fs.statSync(this.dir).isDirectory()) {
      return [];
    }
    return fs
      .readdirSync(this.dir)
      .filter((f) => f.endsWith(".json"))
      .sort()
      .map((f) => path.join(this.dir, f));
  }

  private prune(): void {
    const files = this.files();
    const excess = files.length - MAX_CHECKPOINTS_PER_WORKSPACE;
    for (let i = 0; i < Math.max(excess, 0); i++) {
      const f = files[i];
      if (f) {
        try {
          fs.unlinkSync(f);
        } catch {
          // best-effort
        }
      }
    }
  }

  /** Total number of checkpoints retained on disk (may exceed what list() returns). */
  count(): number {
    return this.files().length;
  }

  /** Most recent first. */
  list(limit = 10): Checkpoint[] {
    if (limit <= 0) {
      return [];
    }
    const files = this.files();
    const tail = files.slice(-limit).reverse();
    const out: Checkpoint[] = [];
    for (const f of tail) {
      try {
        const parsed = checkpointFromJson(JSON.parse(fs.readFileSync(f, "utf-8")));
        if (parsed) {
          out.push(parsed);
        }
      } catch {
        continue;
      }
    }
    return out;
  }

  /** Pop and apply the most recent checkpoint. Returns it, or undefined if empty.
   *
   * Best-effort per file: a problem restoring one change (permission error,
   * a path that no longer makes sense, a symlink now pointing somewhere
   * unexpected) doesn't abort the rest of the checkpoint or throw — every
   * apply below and the checkpoint file itself are always cleaned up.
   */
  undoLast(): Checkpoint | undefined {
    const files = this.files();
    const last = files[files.length - 1];
    if (last === undefined) {
      return undefined;
    }
    let cp: Checkpoint | null;
    try {
      cp = checkpointFromJson(JSON.parse(fs.readFileSync(last, "utf-8")));
    } catch {
      cp = null;
    }
    if (cp === null) {
      this.safeUnlink(last);
      return undefined;
    }
    for (const change of cp.changes) {
      try {
        // Re-resolve at undo time (not just the raw stored path) so a
        // symlink introduced after the checkpoint was recorded can't
        // smuggle the write outside the workspace.
        const target = realpathAsFarAsPossible(change.path);
        const rel = path.relative(this.root, target);
        if (rel === ".." || rel.startsWith(`..${path.sep}`) || path.isAbsolute(rel)) {
          continue; // refuse to touch anything outside this checkpoint's workspace
        }
        if (change.existedBefore) {
          fs.mkdirSync(path.dirname(target), { recursive: true });
          fs.writeFileSync(target, change.beforeContent ?? "", "utf-8");
        } else if (fs.existsSync(target) && fs.statSync(target).isFile()) {
          fs.unlinkSync(target);
        }
      } catch {
        continue; // best-effort: skip this file, still process the rest
      }
    }
    this.safeUnlink(last);
    return cp;
  }

  private safeUnlink(file: string): void {
    try {
      fs.unlinkSync(file);
    } catch {
      // best-effort
    }
  }
}
